engine/sql/core/src/main/scala/org/apache/spark/sql/index/RTree.scala

```diff
@@ -87,6 +87,39 @@ case class RTree(root: RTreeNode) extends Index with Serializable {
     ans.toArray
   }
 
+  def limitedLevelRange(query: MBR, selectivity: Double, level: Int, enable: Boolean = false)
+      : Option[Array[(Shape, Int)]] = {
+    if (!enable) {
+      Some(range(query))
+    } else {
+      require(level >= 0)
+      val ans = mutable.ArrayBuffer[(Shape, Int)]()
+      val st = new mutable.Stack[RTreeNode]()
+      var current_level = 0
+      if (root.m_mbr.intersects(query) && root.m_child.nonEmpty) st.push(root)
+      while (st.nonEmpty) {
+        val now = st.pop()
+        if (current_level <= level && now.m_mbr.intersectRadio(query) > selectivity) {
```

Comment:
You logic here is not correct since you only calculate the ratio for the current R-Tree node. If the predicate only fully covers one node in the second level, you will still start a scan, which is obviously not a good idea.